ExtraDry/ExtraDry.Core/Server/Internal/ModelDescription.cs

```diff
@@ -1,29 +1,89 @@
-﻿using ExtraDry.Core;
+﻿#nullable enable
+
+using ExtraDry.Core;
 using System;
 using System.Collections.ObjectModel;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
 using System.Reflection;
+using System.Text.Json.Serialization;
 
 namespace ExtraDry.Server.Internal {
 
     internal class ModelDescription {
+        private const string DuplicateKeyMessage = "Sort requires that a single EF key is well defined to stabalize the sort, composite keys are not supported.  Manually specify a Stabilizer in the FilterQuery, or use a single [Key] attribute.";
+        private const string MissingStabilizerMessage = "Sort requires that an EF key is uniquely defined to stabalize the sort, even if another sort property is present.  Create a unique key following EF conventions or specify a Stabilizer in the FilterQuery.";
+        private const string UserMessage = "Unable to Sort. {0}";
 
         public ModelDescription(Type modelType)
         {
             GetReflectedModelProperties(modelType);
         }
 
         public Collection<FilterProperty> FilterProperties { get; } = new Collection<FilterProperty>();
+        public Collection<SortProperty> SortProperties { get; } = new Collection<SortProperty>();
+        public SortProperty StabilizerProperty { get; private set; } = null!;
 
         private void GetReflectedModelProperties(Type modelType)
         {
             var properties = modelType.GetProperties();
+            SortProperty? stabilizerPropertyByConvention = default;
             foreach(var property in properties) {
+                var externalName = ExternalName(property);
+
                 var filter = property.GetCustomAttribute<FilterAttribute>();
                 if(filter != null) {
-                    FilterProperties.Add(new FilterProperty(property, filter));
+                    FilterProperties.Add(new FilterProperty(property, filter, externalName));
+                }
+
+                if(IsSortable(property)) {
+                    SortProperties.Add(new SortProperty(property, externalName));
+                }
+
+                var keyProperty = property.GetCustomAttribute<KeyAttribute>();
+                if(keyProperty != default) {
+                    if(StabilizerProperty == default) {
+                        StabilizerProperty = new SortProperty(property, externalName);
+                    }
+                    else {
+                        throw new DryException(DuplicateKeyMessage, string.Format(UserMessage, "0x0F3F241D"));
```

Comment:
Ok, I see a problem with moving the sort property to the ModelDescription.  When done in the Sort method, this exception was thrown in normal imperative logic flow.  However, now it is thrown from a constructor (frowned upon).  Also, what if I have a Model that I'm never going to sort by, I'm still forced to create a single key...

Perhaps the SortProperty should be nullable?  Perhaps it should just be a flag on the properties?  ModelDescription shouldn't care if you have two `KeyAttributes`s  I think the reflection is right to be done here, but not the exceptions, which should move back to the Sort method.

Reply:
I agree there shouldn't be any exceptions thrown from the constructor.  I still like the idea of the `ModelDescription` class of having the responsibility of determining if the model is configured correctly for sorting.  My proposal is to move this logic from `GetReflectedModelProperties` and into the `StabilizerProperty` property i.e.
```C#
public SortProperty StabilizerProperty {
  get {
    if (stabilizerProperty == null) {
      throw new DryException(MissingStabilizerMessage, string.Format(UserMessage, "0x0F3F241C"));
    }
    if (multipleStabilizerProperties) {
      throw new DryException(DuplicateKeyMessage, string.Format(UserMessage, "0x0F3F241D"));
    }
    return stabilizerProperty;
  }
}
```
I think this keep the logic encapsulated and reduces the need for `null` checks on `StabilizerProperty`.

Reply:
There's a similar but weaker argument that properties shouldn't throw exceptions...  The key question is: "Is it valid for a ModelDescription's model to not have a Stabalizer?"

I'm not sure the answer, but an [Owned] object wouldn't have a stabalizer.  Do any of those ever need a ModelDescription? (I haven't looked at the code to be able to answer...)

Reply:
Based on the error message it seems as if there is scope to allow the consumer of the sort functionality to provide a stabilizer property.  Given this is it shouldn't be raising exceptions if it's unable to determine the stabilizer property.

I've removed the raising of exceptions from `ModelDescription` and moved it back to the `Sort` method.